### ISSUE TITLE
training.moe_lora loads on ipo / bco / reward_model / ppo / embedding and none of them reads it (#1074 follow-up)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#1099]
Follow-up from the review of #1074 (ruling (b) on #798).

#1074 wires `training.moe_lora` into sft / pretrain / dpo / kto / orpo / simpo / grpo and refuses the MoE expert knobs where no trainer reads them. `moe_lora: true` itself still **loads** on five more LoRA tasks, and none of those trainers reads it: `ipo`, `bco`, `reward_model`, `ppo`, `embedding`. This was checked by loading such configs on the #1074 merged tree. So on those tasks the flag is accepted and silently ignored, which is the #748 / v0.75.0 defect class.

**Decision to make in this issue before code:** for each of the five, either wire `moe_lora` target selection through the same helper #1074 introduced (`resolve_moe_lora_targets`), or refuse `moe_lora: true` at load for that task with a message naming the task. Wiring is preferred where the trainer builds its adapter through `build_lora_config`, which all of them now do after #1037.

**Before refusing anything, check shipped configs:** no recipe, template or `examples/` config may set `moe_lora: true` on one of these tasks. If one does, the standing rule applies: WARN and ignore this release, refuse next release.

**Acceptance criteria**
- [ ] Each of the five tasks is either wired (a test per trainer asserting the MoE targets reach the adapter, which fails when the helper call is removed) or refused (a parse-time test per task).
- [ ] The #748 consumer guard and `soup doctor --config`'s declared table agree with the outcome.
- [ ] Changelog fragment and a line in `docs/performance-and-quantization.md`.

Depends on #1074 landing first.

**Comment by Srinivasan8888** [comment]
Taking this, as offered on #1074. Held until #1074 lands, as the issue says — but
the decision you asked to make *before* code is measurable now, so here it is with
the measurement rather than the inference.

## Reproduced on `main` (`c1a45efb`), independently of #1074

All five load `moe_lora: true` and none of the five trainers reads it:

```
ipo            LOADS with moe_lora=true
bco            LOADS with moe_lora=true
reward_model   LOADS with moe_lora=true
ppo            LOADS with moe_lora=true
embedding      LOADS with moe_lora=true
```

## The decision: wire all five, refuse none

Your stated preference is wiring "where the trainer builds its adapter through
`build_lora_config`". Checked, rather than assumed — every one of the five does,
and through the same two-call shape as the seven #1074 wires:

| trainer | `resolve_lora_target_modules` | `build_lora_config` |
|---|---|---|
| `ipo.py` | `:253` | `:255` |
| `bco.py` | `:307` | `:309` |
| `reward_model.py` | `:256` | `:258` |
| `ppo.py` | `:541` | `:543` |
| `embedding.py` | `:290` | `:292` |

So there is no split here: it is one `resolve_moe_lora_targets` call per trainer,
the same helper and the same insertion point, and the refusal branch has no member.

## The shipped-config precondition

Swept all 175 recipes plus `src/soup_cli/templates/**` and `examples/**`:

```
shipped configs setting moe_lora on the five tasks: NONE
```

Four of the five tasks *do* appear in the catalogue (`embedding`, `ipo`, `ppo`,
`reward_model`), so the sweep is not passing because the tasks are absent. Nothing
needs the warn-then-refuse staging either way, which is worth recording since it
would have been the decision if one recipe had set it.

## What I will build

Five calls to `resolve_moe_lora_targets(self.model, tcfg, target_modules,
console=console)` at the sites above, and a test per trainer asserting the MoE
targets reach the adapter and naming *its own* trainer — the same shape as
#1074's, so replacing any one call with a no-op fails a test that says which one.
The dropout refusal comes along for free, since it lives in that helper.

Plus the two remaining criteria: the #748 consumer guard and `soup doctor
--config`'s declared table agreeing with the outcome, a changelog fragment, and
the `docs/performance-and-quantization.md` line.

One thing I will check before writing it rather than assume: whether
`embedding.py`'s LoRA block is inside a conditional that a MoE base can skip — it
is the one of the five whose call is indented a level deeper than the others.

**Comment by Srinivasan8888** [comment]
Built and green, and **deliberately not opened as a PR yet** — the issue says this
depends on #1074 landing, and #1074 is where `resolve_moe_lora_targets` comes
from. It opens the moment that merges; nothing here needs your time before then.

Branch: `Srinivasan8888:fix/issue1099-moe-lora-five-trainers` (`c9556102`), based
on #1074's head.

## What it does

One `resolve_moe_lora_targets(self.model, tcfg, target_modules, console)` call per
trainer, at the `resolve_lora_target_modules` site each already had. No new helper
and no refusal branch — as measured in my comment above, all five build through
`build_lora_config`, so the condition you named for preferring the wiring holds
for every one of them.

## The thing I said I would check rather than assume

`embedding.py`'s LoRA block **is** inside a conditional, and it is the right one:
the `else` of `if tcfg.lora.r == 0`. At `r: 0` that trainer full-fine-tunes and
builds no adapter at all (#700), so there is nothing for `moe_lora` to select and
the flag is correctly unreachable there. That is stated in the code comment and in
the docs line, not left for the next reader to rediscover.

## Per-trainer evidence

Removing each call fails exactly one test, and that test names its own trainer:

| removed from | tests failing | id |
|---|---|---|
| `ipo.py` | 1 | `ipo` |
| `bco.py` | 1 | `bco` |
| `reward_model.py` | 1 | `reward_model` |
| `ppo.py` | 1 | `ppo` |
| `embedding.py` | 1 | `embedding` |

Baseline 21 passed, restore green. Two controls per task as well: without the flag
the attach **fails** on a fused-expert MoE (`target_modules: auto` resolves to
nothing, so these tasks could not attach LoRA at all on such a base), and a dense
base adapts identically with the flag on and off.

The tests drive each trainer's own `_setup_transformers` with only the loaders
stubbed — `get_peft_model` and the target resolution are real — against a real
tiny Qwen3-MoE in the head shape each trainer loads
(`Qwen3MoeForCausalLM` / `…ForSequenceClassification` / `Qwen3MoeModel`).

## The other two acceptance criteria

Both are already satisfied, which I would rather say than quietly tick:

- **#748 consumer guard** — `training.moe_lora` is not in `KNOWN_UNCONSUMED`; `sft`
  and `pretrain` already read it, so nothing to retire. The guard passes unchanged.
- **`soup doctor --config`'s declared table** — `backend_support.py:141` records
  the flag as `IGNORED` *on MLX* with `trainer_reads=True`. That is a statement
  about the MLX backend, and it stays true; this PR changes the transformers path.

Docs line added to `docs/performance-and-quantization.md`, including the `r: 0`
caveat. The changelog fragment is the one thing missing, since it needs the PR
number.

Full suite: `1 failed, 23797 passed, 143 skipped` — the failure is
`test_issue580_autodistill_mlx_worker`, which fails identically on clean `main` on
this box (`1 failed, 15 passed` on `2131b683` with nothing of mine applied).

**Comment by Srinivasan8888** [comment]
#1074 has merged, so this is open as **#1148**, rebased on `58c2e3cf`.

Two changes since the branch I described above:

- **`online_dpo` had the same gap.** I flipped `moe_lora: true` on a shipped config for every task before opening. It builds through `build_lora_config` like the five, so it is wired the same way, with its own per-trainer test.
- **`asr` also accepts the flag unread**, but it only trains Whisper, which has no experts, so wiring it would do nothing. I left it as a question in the PR: refuse it at load, or leave it documented.

The control is also now correct in either merge order with #1102. The old one asserted a refusal that #1102 turns into a successful attention-only attach.

**Comment by MakazhanAlpamys** [comment]
Resolved by #1148 (merged as 5f5732ae), by @Srinivasan8888 — thank you.

`training.moe_lora` is now read on all five trainers this issue named (`ipo`, `bco`, `reward_model`, `ppo`, `embedding`), plus a sixth, `online_dpo`, which @Srinivasan8888 found by flipping the flag on a shipped config for every task. Each trainer calls `resolve_moe_lora_targets` at the site where it already resolved its LoRA targets, so the flag reaches `build_lora_config` the same way it does on `sft` / `dpo`, and #798's dropout refusal applies there too. `embedding` wires it only when `lora.r >= 1`; at `r: 0` it full-fine-tunes and has no adapter for the flag to select.

The eight tasks that still accept the flag unread (`classifier`, `reranker`, `cross_encoder`, `distill`, `unlearn`, `asr`, `moe_lora_routing`, `prm`) are tracked in #1151, so I am closing this one as done.

Try: set `training.moe_lora: true` on an `ipo` config with a MoE base; the adapter now lands on the expert projections instead of attention only.
